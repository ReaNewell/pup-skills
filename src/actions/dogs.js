import uuid from 'uuid';
import database from '../firebase/firebase';
import { firebase, storage } from '../firebase/firebase';

// ACTIVATE_DOG
export const activateDog = ( id ) => ({
    type: "ACTIVATE_DOG",
    id
});


// ADD_DOG
export const addDog = ({id, name, breed = "unknown", isActive = false, skills = []}) => ({
    type: "ADD_DOG",
    dog: {
        id,
        name,
        breed,
        isActive,
        skills
    }
});
export const startAddDog = (dogData = {}, picture) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        let currentDog;
        const {
            name = "",
            breed = "",
            isActive = false,
            skills = []
        } = dogData;
        const dog = { name, breed, isActive, skills };

        return database.ref(`users/${uid}/dogs`).push(dog).then((ref) => {
            dispatch(addDog({
                id: ref.key,
                ...dog
            }))
            currentDog = ref.key
        }).then(() => {
            if (picture) {
                const fileName = picture.name;
                const storageRef = storage.ref(`/dogPictures/${uid}/${fileName}`);
                const uploadTask = storageRef.put(picture);
                let downloadURL;

                // Listen for state changes, errors, and completion of the upload.
                uploadTask.on(firebase.storage.TaskEvent.STATE_CHANGED, function(snapshot) {
                // Get task progress, including the number of bytes uploaded and the total number of bytes to be uploaded
                var progress = (snapshot.bytesTransferred / snapshot.totalBytes) * 100;
                console.log('Upload is ' + progress + '% done');
                switch (snapshot.state) {
                    case firebase.storage.TaskState.PAUSED: // or 'paused'
                    console.log('Upload is paused');
                    break;
                    case firebase.storage.TaskState.RUNNING: // or 'running'
                    console.log('Upload is running');
                    break;
                }
                }, function(error) {

                }, function() {
                    // Upload completed successfully, now we can get the download URL
                    console.log('Upload complete.')
                    downloadURL = uploadTask.snapshot.downloadURL;
                    return database.ref(`users/${uid}/dogs/${currentDog}`).update({ 
                        pupPicture: downloadURL,
                        pupPictureName: fileName
                    })
                });
            }
        });
    };
};


// EDIT_DOG
export const editDog = (id, updates) => ({
    type: "EDIT_DOG",
    id,
    updates
});
export const startEditDog = (id, updates, currentPicture, picture) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        const currentPupPicture = currentPicture;
        return database.ref(`users/${uid}/dogs/${id}`).update({
            ...updates
        }).then(() => {
            dispatch(editDog(id, updates));
        }).then(() => {
            if (picture) {
                const fileName = picture.name;
                const storageRef = storage.ref(`/dogPictures/${uid}/${fileName}`);
                const uploadTask = storageRef.put(picture);
                let downloadURL;

                // Listen for state changes, errors, and completion of the upload.
                uploadTask.on(firebase.storage.TaskEvent.STATE_CHANGED, function(snapshot) {
                // Get task progress, including the number of bytes uploaded and the total number of bytes to be uploaded
                var progress = (snapshot.bytesTransferred / snapshot.totalBytes) * 100;
                console.log('Upload is ' + progress + '% done');
                switch (snapshot.state) {
                    case firebase.storage.TaskState.PAUSED: // or 'paused'
                    console.log('Upload is paused');
                    break;
                    case firebase.storage.TaskState.RUNNING: // or 'running'
                    console.log('Upload is running');
                    break;
                }
                }, function(error) {

                }, function() {
                    // Upload completed successfully, now we can get the download URL
                    console.log('Upload complete.')
                    downloadURL = uploadTask.snapshot.downloadURL;
                    return database.ref(`users/${uid}/dogs/${id}`).update({ 
                        pupPicture: downloadURL,
                        pupPictureName: fileName
                    }).then(() => {
                        return storage.ref(`/dogPictures/${uid}/${currentPupPicture}`).delete();
                    });
                });
            }
        });
    }
};


// REMOVE_DOG
export const removeDog = (id) => ({
    type: "REMOVE_DOG",
    id
});
export const startRemoveDog = (id) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/dogs/${id}`).remove().then(() => {
            dispatch(removeDog(id));
        });
    };
};


// SET_DOGS
export const setDogs = (dogs) => ({
    type: "SET_DOGS",
    dogs
});
export const startSetDogs = () => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;

        return database.ref(`users/${uid}/dogs`).once('value').then((snapshot) => {
            const dogs = [];

            snapshot.forEach((childSnapshot) => {
                dogs.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            });

            dispatch(setDogs(dogs));
        });
    };
};

// UPDATE_PICTURE
export const startUpdatePicture = (picture, dogId) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;

        const fileName = picture.name;
        const storageRef = storage.ref(`/dogPictures/${uid}/${fileName}`);
        const uploadTask = storageRef.put(picture);
        let downloadURL;

        // Listen for state changes, errors, and completion of the upload.
        uploadTask.on(firebase.storage.TaskEvent.STATE_CHANGED, function(snapshot) {
        // Get task progress, including the number of bytes uploaded and the total number of bytes to be uploaded
        var progress = (snapshot.bytesTransferred / snapshot.totalBytes) * 100;
        console.log('Upload is ' + progress + '% done');
        switch (snapshot.state) {
            case firebase.storage.TaskState.PAUSED: // or 'paused'
            console.log('Upload is paused');
            break;
            case firebase.storage.TaskState.RUNNING: // or 'running'
            console.log('Upload is running');
            break;
        }
        }, function(error) {

        }, function() {
            // Upload completed successfully, now we can get the download URL
            console.log('Upload complete.')
            downloadURL = uploadTask.snapshot.downloadURL;
            return database.ref(`users/${uid}/dogs/${dogId}`).update({ 
                pupPicture: downloadURL,
                pupPictureName: fileName
            })
        });
    }  
};