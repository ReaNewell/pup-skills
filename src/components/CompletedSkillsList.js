import React from 'react';
import selectSkills from '../selectors/skillsByCategory';


import SkillCard from './SkillCard';

export default class CompletedSkillsList extends React.Component {
    constructor(props) {
        super(props)

        this.state = {
            completedSkills: props.activeDog.skills ? selectSkills(props.activeDog.skills, "Completed") : []
        };
    };
    componentWillReceiveProps(nextProps) {
        this.setState({ completedSkills: selectSkills(nextProps.activeDog.skills, "Completed") });
    };
    render() {
        return (
            <div className={this.props.listIsActive ? "skills-list" : "skills-list--inactive"}>
                <p className="dashboard-skills__list-title">Completed</p>
                {   
                    this.state.completedSkills.length === 0 ? (
                        <p className="skills-list__message">There are no completed skills.</p>
                    ) : (
                        this.state.completedSkills.map((skill, index) => (
                                <SkillCard
                                    key = {skill.id}
                                    skill = {skill}
                                />
                        ))
                    )
                }
            </div>
        )
    }
};