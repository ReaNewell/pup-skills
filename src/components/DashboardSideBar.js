import React from 'react';
import { Link } from 'react-router-dom';

import DogList from './DogList';
import NameCard from './NameCard';

class DashboardSideBar extends React.Component {
    render() {
        return (
            <div className="dashboard-sidebar__container">
                <div className="dashboard-sidebar">
                    <NameCard />
                    <DogList />
                </div>
                <div>
                    <Link to='/settings'>Edit Profile</Link>
                </div>
            </div>
        )
    };
};

export default DashboardSideBar;