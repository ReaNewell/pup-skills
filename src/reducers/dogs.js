const defaultDogsReducerState = [];

export default (state = defaultDogsReducerState, action) => {
    switch (action.type) {
        case "ACTIVATE_DOG":
            return state.map((dog) => {
                if (dog.id === action.id) {
                    return {
                        ...dog,
                        isActive: true
                    }
                } else {
                    return {
                        ...dog,
                        isActive: false
                    };
                }
            });
        case "ADD_DOG":
            const oldState = state.map((dog) => { 
                return { 
                    ...dog, 
                    isActive: false 
                }
            });
            return oldState.concat([{ ...action.dog }]);
        case "SET_DOGS":
            return action.dogs.map((dog) => {
                let skills = [];
                for (var skill in dog.skills) {
                    skills.push({
                        ...dog.skills[skill],
                        id: skill
                    });
                }
                return {
                    ...dog,
                    skills
                }
            });

        case 'ADD_SKILL':
            return state.map((dog) => {
                let skills = [];
                if (dog.skills) {
                    for (let i=0; i<dog.skills.length; i++) {
                        skills[i] = dog.skills[i];
                    }
                }
                return dog.id === action.skill.dogId ? (
                    {
                        ...dog,
                        skills: [
                            ...skills,
                            action.skill
                        ]
                    }
                ) : (
                    dog
                )
            });
        case 'CHANGE_CATEGORY':
            let newCategory = '';
            if (action.category === 'In Progress') {
                newCategory = 'Completed';
            } else {
                newCategory = 'In Progress'
            }
            return state.map((dog) => {
                if (dog.id === action.dogId) {
                    return {
                        ...dog,
                        skills: dog.skills.map((skill) => {
                            if (skill.id === action.id) {
                                return {
                                    ...skill,
                                    category: newCategory
                                } 
                            } else {
                                return skill;       
                            }
                        })
                    }
                } else {
                    return dog;
                }
            });
        case 'REMOVE_SKILL':
            return state.map((dog) => {
                if (dog.id === action.dogId) {
                    return {
                        ...dog,
                        skills: dog.skills.filter(( skill ) => skill.id !== action.id)
                    };
                } else {
                    return dog;
                }
            });

        default:
            return state;
    }
};