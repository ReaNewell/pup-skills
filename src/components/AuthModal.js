import React from 'react';
import { connect } from 'react-redux';
import { startLogin, startLoginWithEmail, startSignUp } from '../actions/auth';

class AuthModal extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            email: "",
            error: null,
            password: ""
        }
    }
    onEmailChange = (e) => {
        const email = e.target.value;
        this.setState(() => ({ email }));
    }
    onPasswordChange = (e) => {
        const password = e.target.value;
        this.setState(() => ({ password }));
    }
    loginWithEmail = () => {
        if (this.state.email && this.state.password) {
            this.props.startLoginWithEmail(this.state.email, this.state.password);
        } else {
            this.setState(() => ({ error: "The email and/or password is invalid." }))
        }
    }
    signupWithEmail = () => {
        if (this.state.email && /^(?=.*[A-Za-z])(?=.*\d)[A-Za-z\d]{8,}$/.test(this.state.password)) {
            this.props.startSignUp(this.state.email, this.state.password);
        } else {
            this.setState(() => ({ error: "The email and/or password is invalid. Password must contain eight characters, at least one number, and at least one letter." }))
        }
    }
    render() {
        return (
            <div className="auth-modal">
                <div className="auth-modal__box">
                    {this.state.error && <p className='auth-modal__error'>{this.state.error}</p>}
                    <p className='auth-modal__exit' onClick={this.props.closeLoginModal}>Nevermind.</p>
                    <input 
                        className="auth-modal__email"
                        onChange={this.onEmailChange}
                        placeholder='Email'
                        type='email'
                        value={this.state.email}
                    />
                    <input 
                        className="auth-modal__password"
                        onChange={this.onPasswordChange}
                        placeholder='Password'
                        type='password'
                        value={this.state.password}
                    />
                    <button onClick={this.signupWithEmail} className="auth-modal__button--signUp">Sign Up</button>
                    <button onClick={this.loginWithEmail} className="auth-modal__button--signIn">Login</button>
                </div>
            </div>
        )
    };
};

const mapStateToProps = (state) => ({
    isAuthenticated: !!state.auth.uid
})

const mapDispatchToProps = (dispatch) => ({
    startLogin: () => dispatch(startLogin()),
    startSignUp: (email, password) => dispatch(startSignUp(email, password)),
    startLoginWithEmail: (email, password) => dispatch(startLoginWithEmail(email, password))
});

export default connect(mapStateToProps, mapDispatchToProps)(AuthModal);