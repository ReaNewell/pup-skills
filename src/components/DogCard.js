import React from 'react';
import { connect } from 'react-redux';
import { activateDog } from '../actions/dogs';

export class DogCard extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            id: props.dog.id,
            name: props.dog.name,
            picture: props.dog.pupPicture
        }
    };
    // Deactivates all dogs, activates Dog Card that is clicked.
    activateDog = () => {
        this.props.activateDog(this.state.id);
        this.props.toggleBar && this.props.toggleBar();
    };
    render () {
        return (
            <div className={this.props.dog.isActive ? "dog-card dog-card--active" : "dog-card"} onClick={this.activateDog}>
                {this.state.picture && <img src={this.state.picture} className='dog-card__picture'/>}
                <p className='dog-card__name'>{this.state.name}</p>
            </div>
        );
    };
};

const mapDispatchToProps = (dispatch) => ({
    activateDog: (data) => dispatch(activateDog(data))
});

export default connect(undefined, mapDispatchToProps)(DogCard);