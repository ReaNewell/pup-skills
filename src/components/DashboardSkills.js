import React from 'react';
import { connect } from 'react-redux';
import activeDogSelector from '../selectors/activeDog';

import CompletedSkillsList from './CompletedSkillsList';
import InProgressSkillsList from './InProgressSkillsList';
import SkillCard from './SkillCard';
import SkillsForm from './SkillsForm';

class DashboardSkills extends React.Component {
    constructor(props) {
        super(props)

        this.state = {
            dogs: props.dogs,
            activeDog: props.activeDog,
            completedListIsActive: true,
            skillModalOpen: false
        }
    }
    componentWillReceiveProps(nextProps) {
        this.setState({ dogs: nextProps.dogs, });
        this.setState({ activeDog: nextProps.activeDog });
    };

    openSkillModal = () => {
        this.setState(() => ({ skillModalOpen: true }));
    }
    closeSkillModal = () => {
        this.setState(() => ({ skillModalOpen: false }));
    }
    render() {
        return (
            <div className="dashboard-skills">
                {
                    this.state.activeDog && 
                    <div className="dashboard-skills__container">
                        <div className="dashboard-skills__header">
                            <h1 className="dashboard-skills__title">Here are {this.state.activeDog.name}'s skills</h1>
                            <button className="dashboard-skills__button" onClick={this.openSkillModal}>Add Skill</button>
                        </div>
                        {this.state.skillModalOpen && <SkillsForm closeSkillModal={this.closeSkillModal} activeDog={this.state.activeDog}/>}
                        <div className="dashboard-skills__lists">
                            <CompletedSkillsList listIsActive={this.state.completedListIsActive} activeDog={this.state.activeDog}/>
                            <InProgressSkillsList listIsActive={this.state.completedListIsActive} activeDog={this.state.activeDog}/>
                        </div>
                    </div>
                }
            </div>
        );
    };
};

const mapStateToProps = (state) => {
    return { 
        dogs: state.dogs,
        activeDog: activeDogSelector(state.dogs)
    };
};
export default connect(mapStateToProps)(DashboardSkills);