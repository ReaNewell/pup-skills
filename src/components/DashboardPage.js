import React from 'react';
import {connect} from 'react-redux';

import DashboardSkills from './DashboardSkills';
import DashboardSideBar from './DashboardSideBar';
import MobileDashboardSideBar from './mobile/MobileDashboardSideBar';
import MobileDashboardSkills from './mobile/MobileDashboardSkills';

export class DashboardPage extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            width: window.innerWidth
        }
    };
    componentWillMount() {
        window.addEventListener('resize', this.handleWindowSizeChange);
        if (!this.props.hasProfileName) {
            this.props.history.push('/getting-started');
        }
    };
    componentWillUnmount() {
        window.removeEventListener('resize', this.handleWindowSizeChange);
    };
    handleWindowSizeChange = () => {
        this.setState({ width: window.innerWidth });
    };
    render() {
        return (
            <div className='dashboard-page__background'>
                <div className="dashboard-page content-container">
                    {this.state.width > 720 ? <DashboardSideBar /> : <MobileDashboardSideBar/>}
                    {this.state.width > 720 ? <DashboardSkills /> : <MobileDashboardSkills/>}
                </div>
            </div>
        )
    }
};

const mapStateToProps = (state) => ({
    hasProfileName: !!state.profileInfo.profileName
});
export default connect(mapStateToProps)(DashboardPage);