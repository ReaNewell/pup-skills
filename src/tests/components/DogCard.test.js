import React from 'react';
import { shallow } from 'enzyme';
import { DogCard } from '../../components/DogCard';
import dogs from '../fixtures/dogs';

test('should render DogCard component correctly', () => {
    const wrapper = shallow(<DogCard dog={dogs[0]} />);
    expect(wrapper).toMatchSnapshot();
});

test('should call activateDog when component is clicked', () => {
    const activateDog = jest.fn();
    const wrapper = shallow(<DogCard dog={dogs[0]} activateDog={activateDog} />);
    wrapper.simulate('click');
    expect(activateDog).toHaveBeenCalled();
});