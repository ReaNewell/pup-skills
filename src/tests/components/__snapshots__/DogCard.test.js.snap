// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render DogCard component correctly 1`] = `
<div
  className="dog-card dog-card--active"
  onClick={[Function]}
>
  <p
    className="dog-card__name"
  >
    Pheobe
  </p>
</div>
`;
